### PR TITLE
Implement compiler support for dictionary

### PR DESCRIPTION
**Pull request by sklam**
(Based on #3775, #3760, #3715) 

Implement most dictionary operations.

Still missing:

- ~~comparisons (equality)~~
- boxing and unboxing
- cpython side interface
- refct types are not supported; only works on POD types.
- typeinference backtracking magic
- (am I missing anything?)

Actual changes starts at https://github.com/numba/numba/pull/3777/commits/c56c5ac5795166dfdc3835dccf4019c127ca10ed.

The major changes are in:
- numba/dictobject.py  for the compiler support of the dictionary
- numba/types/containers.py for the dictionary typing
- numba/tests/test_dictobject.py for the tests

Note: `TypeRef` is added to allow referencing any numba-type as global-value or arguments as used in the constructor of dictionary.  It is different from `NumberClass`, which is for i.e. `np.float32`.

### CODE REVIEW

**Comment by sklam** [comment]
There are 2 unrelated failures.
- linux py35_np111_cov is failing due to network connection error.
- windows py27 is failing due to DLL hell.

File: numba/dictobject.py [hunk 1]
```diff
@@ -0,0 +1,922 @@
...
+    sig = casted(casted)
+    return sig, codegen
+
+
+@overload(operator.setitem)
+def impl_setitem(d, key, value):
+    if not isinstance(d, types.DictType):
+        return
+
+    keyty, valty = d.key_type, d.value_type
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/dictobject.py]: Deferring this to PR implementing support of refct'ed type

File: numba/dictobject.py [hunk 2]
```diff
@@ -0,0 +1,922 @@
...
+
+    def impl(d, key, value):
+        key = _cast(key, keyty)
+        val = _cast(value, valty)
+        status = _dict_insert(d, key, hash(key), val)
+        if status == 0:
+            return
+        elif status == 1:
+            # replaced
+            # XXX handle refcount
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/dictobject.py]: to be fixed in a later PR

File: numba/dictobject.py [hunk 3]
```diff
@@ -0,0 +1,922 @@
+
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/dictobject.py]: added docstring there

File: numba/dictobject.py [hunk 4]
```diff
@@ -0,0 +1,922 @@
...
+ll_dictiter_type = cgutils.voidptr_t
+ll_status = cgutils.int32_t
+ll_ssize_t = cgutils.intp_t
+ll_hash = ll_ssize_t
+ll_bytes = cgutils.voidptr_t
+
+# The following enums must match _dictobject.c
+DKIX_EMPTY = -1
+
+OK_REPLACED = 1
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/dictobject.py]: many `ERR_*` are unused for now.   btw, i have converted them to enums

File: numba/dictobject.py [hunk 5]
```diff
@@ -0,0 +1,922 @@
...
+from numba import cgutils
+from numba import _helperlib
+
+from numba.extending import (
+    overload,
+    overload_method,
+    intrinsic,
+    register_model,
+    models,
+    make_attribute_wrapper,
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/dictobject.py]: deleted

File: numba/dictobject.py [hunk 6]
```diff
@@ -0,0 +1,922 @@
...
+
+        dstruct.meminfo = meminfo
+
+        return dstruct._getvalue()
+
+    sig = dict_ty(keyty, valty, ptr)
+    return sig, codegen
+
+
+def _dict_get_data(context, builder, dict_ty, d):
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/dictobject.py]: done

File: numba/dictobject.py [hunk 7]
```diff
@@ -0,0 +1,922 @@
...
+        [td] = sig.args
+        dp = _dict_get_data(context, builder, td, d)
+        n = builder.call(fn, [dp])
+        return n
+
+    return sig, codegen
+
+
+@intrinsic
+def _dict_dump_keys(typingctx, d):
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/dictobject.py]: renamed

File: numba/dictobject.py [hunk 8]
```diff
@@ -0,0 +1,922 @@
...
+            [
+                dp,
+                _as_bytes(builder, ptr_key),
+                _as_bytes(builder, ptr_val),
+            ],
+        )
+        out = context.make_optional_none(builder, keyvalty)
+        pout = cgutils.alloca_once_value(builder, out)
+
+        cond = builder.icmp_signed('==', status, status.type(0))
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/dictobject.py]: fixed

File: numba/dictobject.py [hunk 9]
```diff
@@ -0,0 +1,922 @@
...
+@overload(new_dict)
+def impl_new_dict(key, value):
+    """Creates a new dictionary with *key* and *value* as the type
+    of the dictionary key and value, respectively.
+    """
+    if any([
+        not isinstance(key, Type),
+        not isinstance(value, Type),
+    ]):
+        raise TypeError
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/dictobject.py]: added a msg

File: numba/dictobject.py [hunk 10]
```diff
@@ -0,0 +1,922 @@
...
+        [val, typ] = args
+        return val
+    # Using implicit casting in argument types
+    casted = typ.instance_type
+    sig = casted(casted, typ)
+    return sig, codegen
+
+
+@intrinsic
+def _nonoptional(typingctx, val):
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/dictobject.py]: TBD: https://github.com/numba/numba/issues/3797

File: numba/dictobject.py [hunk 11]
```diff
@@ -0,0 +1,922 @@
...
+            return
+        elif status == ERR_CMP_FAILED:
+            raise ValueError('key comparison failed')
+        else:
+            raise RuntimeError('dict.__setitem__ failed unexpectedly')
+    return impl
+
+
+@overload_method(types.DictType, 'get')
+def impl_get(dct, k, d=None):
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/dictobject.py]: switched to `default`

File: numba/dictobject.py [hunk 12]
```diff
@@ -0,0 +1,921 @@
...
+
+
+@overload_method(types.DictType, 'popitem')
+def impl_popitem(d):
+    if not isinstance(d, types.DictType):
+        return
+
+    def impl(d):
+        status, keyval = _dict_popitem(d)
+        if status == 0:
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/dictobject.py]: fixed

**Comment by sklam** [comment]
https://github.com/numba/numba/pull/3777/commits/d5befda4842a9f6994cecde595630c99de8213cb added a unsafe casting check in `_cast`

**Comment by sklam** [comment]
Included tests from @stuartarchibald.  
3 tests are disabled:
- 2 tests are using unicode_type that needs refct.  They will be handled in a later PR.
- 1 test is disabled due to lack of support for using `dict.keys()` as a set-like object.

File: numba/dictobject.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/dictobject.py]: ok, sounds good.

**Comment by stuartarchibald** [comment]
Thanks for all the work on this @sklam, it seems stable as a ground work PR for the python side of this feature. Testing covers the functionality presented and the implementation seems to match the definition, marking for merge!

**Pull request merged**